jupyterhub_traefik_proxy/toml.py

```diff
@@ -0,0 +1,300 @@
+"""Traefik implementation
+
+Custom proxy implementations can subclass :class:`Proxy`
+and register in JupyterHub config:
+
+.. sourcecode:: python
+
+    from mymodule import MyProxy
+    c.JupyterHub.proxy_class = MyProxy
+
+Route Specification:
+
+- A routespec is a URL prefix ([host]/path/), e.g.
+  'host.tld/path/' for host-based routing or '/path/' for default routing.
+- Route paths should be normalized to always start and end with '/'
+"""
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+import json
+import os
+import asyncio
+
+from urllib.parse import urlparse
+from traitlets import Any, default, Unicode
+
+from . import traefik_utils
+from jupyterhub.proxy import Proxy
+from jupyterhub_traefik_proxy import TraefikProxy
+
+
+class TraefikTomlProxy(TraefikProxy):
+    """JupyterHub Proxy implementation using traefik and toml config file"""
+
+    mutex = Any()
+
+    @default("mutex")
+    def _default_mutex(self):
+        return asyncio.Lock()
+
+    toml_static_config_file = Unicode(
+        "traefik.toml", config=True, help="""traefik's static configuration file"""
+    )
+
+    toml_dynamic_config_file = Unicode(
+        "rules.toml", config=True, help="""traefik's dynamic configuration file"""
+    )
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.routes_cache = {}
+
+    async def _setup_traefik_static_config(self):
+        self.log.info("Setting up traefik's static config...")
+        self._generate_htpassword()
+
+        try:
+            with open(self.toml_static_config_file, "w") as f:
+                f.write('defaultEntryPoints = ["http"]\n')
```

Comment:
Instead of writing toml-as-text, let's use the [toml package](https://github.com/uiri/toml) to ensure we always create valid toml.